src/StackExchange.Redis/CommandRetry/MessageRetryQueue.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace StackExchange.Redis
+{
+    internal class MessageRetryQueue : IDisposable
+    {
+        private readonly Queue<Message> _queue = new Queue<Message>();
+        private readonly IMessageRetryHelper _messageRetryHelper;
+        private readonly int? _maxRetryQueueLength;
+        private readonly bool _runRetryLoopAsync;
+
+        internal MessageRetryQueue(IMessageRetryHelper messageRetryHelper, int? maxRetryQueueLength = null, bool runRetryLoopAsync = true)
+        {
+            _maxRetryQueueLength = maxRetryQueueLength;
+            _runRetryLoopAsync = runRetryLoopAsync;
+            _messageRetryHelper = messageRetryHelper;
+        }
+
+        public int CurrentRetryQueueLength => _queue.Count;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal bool TryHandleFailedCommand(Message message)
+        {
+            bool wasEmpty;
+            lock (_queue)
+            {
+                int count = _queue.Count;
+                if (_maxRetryQueueLength.HasValue && count >= _maxRetryQueueLength)
+                {
+                    return false;
+                }
+                wasEmpty = count == 0;
+                _queue.Enqueue(message);
+            }
+            if (wasEmpty) StartRetryQueueProcessor();
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void StartRetryQueueProcessor()
+        {
+            bool startProcessor = false;
+            lock (_queue)
+            {
+                startProcessor = _queue.Count > 0;
+            }
+            if (startProcessor)
+            {
+                if (_runRetryLoopAsync)
+                {
+                    var task = Task.Run(ProcessRetryQueueAsync);
```

Comment:
Based on our last experience here, need to switch this to thread kickoff /cc @mgravell @deepakverma